### PR TITLE
Add more typing to transaction and base client

### PR DESCRIPTION
**Pull request by jameshilliard**
Also fix some issues identified by additional checks.

### CODE REVIEW

**Review by janiversen — COMMENTED** [review]
Changes look good, but there are a couple of CI problems.

**Comment by jameshilliard** [comment]
> there are a couple of CI problems

Hmm, fixed everything except a test coverage percentage check, not sure what to do about that one.

**Comment by jameshilliard** [comment]
Tests passing now.

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -98,7 +98,7 @@ jobs:
       - name: ruff
         if: matrix.run_lint == true
         run: |
-          ruff .
+          ruff check .
```
Comment (janiversen) [inline .github/workflows/ci.yml:101]: We do not mix CI changes with other changes, simply because changing CI requires a bit more than just a PR.

File: pyproject.toml [hunk 2]
```diff
@@ -240,7 +240,7 @@ exclude_also = [
     "if __name__ == .__main__.:",
     ]
 skip_covered = true
-fail_under = 90.0
+fail_under = 89.0
```
Comment (janiversen) [inline pyproject.toml:243]: A not acceptable change, we want to have a higher % not lower !!

**Review by janiversen — CHANGES_REQUESTED** [review]

File: pyproject.toml [hunk 3]
```diff
@@ -62,17 +62,17 @@ documentation = [
 development = [
     "build>=1.2.1",
     "codespell>=2.3.0",
-    "coverage>=7.5.3",
-    "mypy>=1.10.0",
-    "pylint>=3.2.3",
+    "coverage>=7.6.0",
```
Comment (janiversen) [inline pyproject.toml:65]: Bump 3party versions needs to be in a seperate PR; only with the bump.

**Review by janiversen — CHANGES_REQUESTED** [review]

**Comment by janiversen** [comment]
And just for the fun of it, I run into lowering the % in #2239, but instead of lowering the bar, I will examine the cause and add tests.

File: pyproject.toml (hunk 3 above)
Comment (jameshilliard) [reply pyproject.toml]: Moved to #2241.

File: pyproject.toml (hunk 2 above)
Comment (jameshilliard) [reply pyproject.toml:243]: Hmm, since this code is going to be [overhauled anyways](https://github.com/pymodbus-dev/pymodbus/pull/2232#pullrequestreview-2176716262) does test coverage percentage matter all that much right now? Getting typing working better should hopefully make it easier to identify unused codepaths and do further cleanup so I'm not sure we should be actually be all that concerned about coverage at the moment for things that are being refactored like the transaction code. I'm suspecting that there's some unused code branches in the sync client transaction code.

File: .github/workflows/ci.yml (hunk 1 above)
Comment (jameshilliard) [reply .github/workflows/ci.yml:101]: Hmm, tests don't pass without this change, btw I also added this to the dev deps bump PR in #2241 since it is needed there as well I think for tests to pass.

File: pymodbus/transaction.py [hunk 4]
```diff
@@ -368,16 +374,13 @@ def _transact(self, request: ModbusRequest, response_length, full=False, broadca
             packet = self.client.framer.buildPacket(request)
             Log.debug("SEND: {}", packet, ":hex")
             size = self._send(packet)
-            if (
-                isinstance(size, bytes)
-                and self.client.state == ModbusTransactionState.RETRYING
-            ):
+            if self.client.state == ModbusTransactionState.RETRYING:
```
Comment (janiversen) [inline pymodbus/transaction.py:377]: This change do not look correct, we only want to enter the "if" if we received something.

This have nothing to do with typing, and it is not covered by tests.

File: pymodbus/transaction.py [hunk 5]
```diff
@@ -368,16 +374,13 @@ def _transact(self, request: ModbusRequest, response_length, full=False, broadca
...
                 Log.debug(
                     "Changing transaction state from "
                     '"RETRYING" to "PROCESSING REPLY"'
                 )
                 self.client.state = ModbusTransactionState.PROCESSING_REPLY
-                return size, None
+                return b"", None
```
Comment (janiversen) [inline pymodbus/transaction.py:383]: This is wrong, we only change to REPLY when we have a reply.

This have nothing to do with typing, and it is not covered by tests.

File: pymodbus/transaction.py [hunk 6]
```diff
@@ -420,8 +423,10 @@ def _recv(self, expected_response_length, full) -> bytes:  # noqa: C901
                 min_size = 4
             elif isinstance(self.client.framer, ModbusAsciiFramer):
                 min_size = 5
-            else:
+            elif expected_response_length:
```
Comment (janiversen) [inline pymodbus/transaction.py:426]: This do not seem right, however it is not covered by tests.

File: pymodbus/transaction.py [hunk 7]
```diff
@@ -463,13 +468,17 @@ def _recv(self, expected_response_length, full) -> bytes:  # noqa: C901
                         expected_response_length -= min_size
                         total = expected_response_length + min_size
                 else:
+                    if exception_length is None:
```
Comment (janiversen) [inline pymodbus/transaction.py:471]: Not covered by test.

File: pymodbus/transaction.py [hunk 8]
```diff
@@ -463,13 +468,17 @@ def _recv(self, expected_response_length, full) -> bytes:  # noqa: C901
...
+                        exception_length = 0
                     expected_response_length = exception_length - min_size
                     total = expected_response_length + min_size
             else:
                 total = expected_response_length
         else:
             read_min = b""
             total = expected_response_length
+        if expected_response_length is None:
```
Comment (janiversen) [inline pymodbus/transaction.py:480]: Not covered by test.

**Review by janiversen — CHANGES_REQUESTED** [review]
May I politely suggest, you do typing in this PR and not a lot of code changes !

The typing in this PR is OK, the code changes are not covered by tests, and a number of the changes seems wrong.

File: pymodbus/transaction.py (hunk 4 above)
Comment (jameshilliard) [reply pymodbus/transaction.py:377]: I removed the `isinstance(size, bytes)` condition, which based on the return type of `self._send(packet)` is always an integer. This was flagged by the type checker.

File: pymodbus/transaction.py (hunk 5 above)
Comment (jameshilliard) [reply pymodbus/transaction.py:383]: > This is wrong, we only change to REPLY when we have a reply.

The issue here is that we can't return an integer, the calling code expects bytes to be returned.

File: pymodbus/transaction.py (hunk 6 above)
Comment (jameshilliard) [reply pymodbus/transaction.py:426]: This was flagged by the type-checker, if `expected_response_length` is none we can't assign `min_size` to `expected_response_length` as we pass `min_size` to `self.client.framer.recvPacket` which expects an integer.

File: pymodbus/transaction.py (hunk 7 above)
Comment (jameshilliard) [reply pymodbus/transaction.py:471]: This was flagged by the type-checker, if `exception_length` is `None` we would try to subtract from `exception_length` in `expected_response_length = exception_length - min_size` which would not work.

File: pymodbus/transaction.py (hunk 8 above)
Comment (jameshilliard) [reply pymodbus/transaction.py:480]: This was flagged by the type-checker, `expected_response_length` can't be `None` as we pass `expected_response_length` to `self.client.framer.recvPacket` which expects an integer.

File: pyproject.toml (hunk 2 above)
Comment (jameshilliard) [reply pyproject.toml:243]: > You do not need to “suspect”, use “pytest —cov” and you know.

I mean, the code additions were all entirely to fix type check errors, I have no idea if some of those branches in the sync client transaction code can even be hit in practice.

> Exactly refactoring code is a good reason to have near 100% coverage before starting, because it helps detect side effects.

Ideally but seems main problem is the type checker is flagging a bunch of issues(lack of checks for `None` for example) and I wouldn't know how to write tests to hit those conditions, the sync transaction code is super messy and hard to follow, at least with type checks we can start restricting the allowable types in various args/variables more easily.

**Comment by jameshilliard** [comment]
> May I politely suggest, you do typing in this PR and not a lot of code changes !

All code changes made are directly due to issues being flagged by the typing additions.

**Comment by janiversen** [comment]
Well some of them are wrong.

**Comment by janiversen** [comment]
Please see my review comments, and when done request a new review.

File: pymodbus/transaction.py (hunk 5 above)
Comment (janiversen) [reply pymodbus/transaction.py]: The code as it is works or not ?

In any case a test case would show the current code to be faulty, and your new code to be correct.

File: pymodbus/transaction.py (hunk 6 above)
Comment (janiversen) [reply pymodbus/transaction.py]: Seems you need more typing.

File: pymodbus/transaction.py (hunk 7 above)
Comment (janiversen) [reply pymodbus/transaction.py]: What has that to do with that you have not added test cases, that show why the old code was broken and the new code works.

File: pymodbus/transaction.py (hunk 8 above)
Comment (janiversen) [reply pymodbus/transaction.py]: Did you read my comment ?

File: .github/workflows/ci.yml (hunk 1 above)
Comment (alexrudd2) [reply .github/workflows/ci.yml]: https://github.com/pymodbus-dev/pymodbus/pull/2241#issuecomment-2239907728

File: pymodbus/client/base.py [hunk 9]
```diff
@@ -4,9 +4,9 @@
 import asyncio
 import socket
 from abc import abstractmethod
-from collections.abc import Awaitable, Callable
+from collections.abc import Callable, Coroutine
 from dataclasses import dataclass
-from typing import Any, cast
+from typing import Any, Union, cast
```
Comment (alexrudd2) [inline pymodbus/client/base.py:9]: Use `from __future__ import annotations` to match the rest of the codebase.

This allows using `X | Y` for instead of `Union[X | Y]` [Union types ](https://docs.python.org/3/library/stdtypes.html#union-type)

**Review by alexrudd2 — CHANGES_REQUESTED** [review]
I am commenting only on the use of `from typing import Union`.

Although I think the other typing changes are worthwhile (in particular the existing code does not perfectly handle `None` or `ModbusException`), when I last looked at them I was unable to find a clean solution.

File: pymodbus/client/base.py (hunk 9 above)
Comment (jameshilliard) [reply pymodbus/client/base.py:9]: > Use `from __future__ import annotations` to match the rest of the codebase.

https://github.com/pymodbus-dev/pymodbus/blob/72e3399d41c20719ac3e63412fd1be2b8cca3840/pymodbus/client/base.py#L2

That import already exists.

> This allows using `X | Y` for instead of `Union[X | Y]` [Union types ](https://docs.python.org/3/library/stdtypes.html#union-type)

Not sure why exactly, but the normal style doesn't seem to work here.

Comment (alexrudd2) [reply pymodbus/client/base.py:9]: ??

Works for me with both `pyright` and `mypy`.  Did you mismatch a bracket, perhaps?

image

File: pyproject.toml (hunk 2 above)
Comment (janiversen) [reply pyproject.toml]: Well maybe you have defined the wrong types, which at least seems to be a possibility.

If you change code, we need to verify why it (in your opinion) did not work earlier, and why it works now. I strongly believe that it is a sound and natural demand ! we do not do changes just because the typing you define causes problems.

That all being said, there might of  and most likely are bugs in the code, which proper typing will bring to the surface, but I am sure you understand, why proof is needed..

If it's that hard to write tests, then maybe it is also hard to write the correct typing, and not just guessing what is correct !! meaning you could always reduce the PR to the part where you know how to write tests.

Adding types to code you do not understand, it quite a mountain to climb !! it is hard enough to add typing to code you understand. Just to give you an example, we had discussions during month how to add typing to mixin.py until someone suggested templates.

File: pymodbus/transaction.py (hunk 4 above)
Comment (janiversen) [reply pymodbus/transaction.py]: But it still needs to get argumented and tested. I see quite a different behavior, because you return a fixed b"", which basically never should be returned in a retry situation.

File: pymodbus/transaction.py (hunk 8 above)
Comment (janiversen) [reply pymodbus/transaction.py]: None can be converted to a 0 without problems.

File: pymodbus/client/base.py (hunk 9 above)
Comment (jameshilliard) [reply pymodbus/client/base.py]: > Works for me with both `pyright` and `mypy`. Did you mismatch a bracket, perhaps?

Oh, the error is coming from `pylint`, not `pyright` or `mypy`:
```python
$ pylint --recursive=y examples pymodbus test
************* Module pymodbus.client.base
pymodbus/client/base.py:23:61: E1131: unsupported operand type(s) for | (unsupported-binary-operation)
pymodbus/client/base.py:228:45: E1131: unsupported operand type(s) for | (unsupported-binary-operation)
```

File: pymodbus/transaction.py (hunk 5 above)
Comment (jameshilliard) [reply pymodbus/transaction.py]: > The code as it is works or not ?

I guess? I mean the tests seem to pass either way so it's a bit hard to tell.

I went ahead and removed this return entirely which also seems to work.

File: pymodbus/transaction.py (hunk 6 above)
Comment (jameshilliard) [reply pymodbus/transaction.py]: Apparently tweaking the style here a bit using original behavior got it passing the type checker.

File: pymodbus/transaction.py (hunk 7 above)
Comment (jameshilliard) [reply pymodbus/transaction.py]: > What has that to do with that you have not added test cases, that show why the old code was broken and the new code works.

Tweaked this a bit more to get it passing the type checker a slightly different way, I'm not sure how we would actually hit this codepath however so a bit hard to write tests.

File: pymodbus/transaction.py (hunk 8 above)
Comment (jameshilliard) [reply pymodbus/transaction.py]: > None can be converted to a 0 without problems.

Updated to no longer convert to 0.

File: pymodbus/client/base.py (hunk 9 above)
Comment (alexrudd2) [reply pymodbus/client/base.py:9]: Bah, appears to be a `pylint` bug.  https://github.com/pylint-dev/pylint/issues/7381.

Use `# pylint: disable=unsupported-binary-operation`, I guess.

File: pymodbus/transaction.py (hunk 4 above)
Comment (jameshilliard) [reply pymodbus/transaction.py]: > I see quite a different behavior, because you return a fixed b"", which basically never should be returned in a retry situation.

I removed the return entirely and changed this check to verify returned size is not 0.

File: pymodbus/client/base.py (hunk 9 above)
Comment (jameshilliard) [reply pymodbus/client/base.py]: Hmm, the normal `|` style seems to be entirely broken on python 3.9:
```python
ERROR test/framers/test_old_framers.py - TypeError: unsupported operand type(s) for |: 'type' and 'NoneType'
ERROR test/framers/test_old_framers.py - TypeError: unsupported operand type(s) for |: 'type' and 'NoneType'
ERROR test/framers/test_old_framers.py - TypeError: unsupported operand type(s) for |: 'type' and 'NoneType'
ERROR test/sub_client/test_client.py - TypeError: unsupported operand type(s) for |: 'type' and 'NoneType'
ERROR test/sub_client/test_client.py - TypeError: unsupported operand type(s) for |: 'type' and 'NoneType'
ERROR test/sub_client/test_client.py - TypeError: unsupported operand type(s) for |: 'type' and 'NoneType'
ERROR test/sub_client/test_client_sync.py - TypeError: unsupported operand type(s) for |: 'type' and 'NoneType'
ERROR test/sub_client/test_client_sync.py - TypeError: unsupported operand type(s) for |: 'type' and 'NoneType'
ERROR test/sub_client/test_client_sync.py - TypeError: unsupported operand type(s) for |: 'type' and 'NoneType'
ERROR test/sub_examples/test_client_server_async.py - TypeError: unsupported operand type(s) for |: 'type' and 'NoneType'
ERROR test/sub_examples/test_client_server_async.py - TypeError: unsupported operand type(s) for |: 'type' and 'NoneType'
ERROR test/sub_examples/test_client_server_async.py - TypeError: unsupported operand type(s) for |: 'type' and 'NoneType'
ERROR test/sub_examples/test_client_server_sync.py - TypeError: unsupported operand type(s) for |: 'type' and 'NoneType'
ERROR test/sub_examples/test_client_server_sync.py - TypeError: unsupported operand type(s) for |: 'type' and 'NoneType'
ERROR test/sub_examples/test_client_server_sync.py - TypeError: unsupported operand type(s) for |: 'type' and 'NoneType'
ERROR test/sub_examples/test_examples.py - TypeError: unsupported operand type(s) for |: 'type' and 'NoneType'
ERROR test/sub_examples/test_examples.py - TypeError: unsupported operand type(s) for |: 'type' and 'NoneType'
ERROR test/sub_examples/test_examples.py - TypeError: unsupported operand type(s) for |: 'type' and 'NoneType'
ERROR test/test_network.py - TypeError: unsupported operand type(s) for |: 'type' and 'NoneType'
ERROR test/test_network.py - TypeError: unsupported operand type(s) for |: 'type' and 'NoneType'
ERROR test/test_network.py - TypeError: unsupported operand type(s) for |: 'type' and 'NoneType'
```

Comment (alexrudd2) [reply pymodbus/client/base.py:9]: To date, the test have not had type hints (since it seemed like low value).  No problem with adding them, but in this case you will need to add `from __future__ import annotations` for the first time.

Comment (jameshilliard) [reply pymodbus/client/base.py]: > No problem with adding them, but in this case you will need to add `from __future__ import annotations` for the first time.

Tried this, doesn't seem to help, looks to just be a limitation of python 3.9.

Comment (janiversen) [reply pymodbus/client/base.py:9]: I agree with @alexrudd2 adding typing to tests, hav3 no real value and will most likely just complicate writing tests.

Please do not go down that route.

**Comment by janiversen** [comment]
Please stop extending this PR, you still have not solved the basic problems in transaction.py

Adding test files to the mix, will only ensure that you will not ge5 the PR reviewed !

File: pymodbus/client/base.py (hunk 9 above)
Comment (jameshilliard) [reply pymodbus/client/base.py:9]: > adding typing to tests, hav3 no real value and will most likely just complicate writing tests.

So I'm not really trying to add types to the tests so much as make it so that we don't get type errors when we typecheck the tests.

**Comment by jameshilliard** [comment]
> Please stop extending this PR, you still have not solved the basic problems in transaction.py

Yeah, I should probably split off those test fixes, it ended up being more fixes than I expected. Was trying to fix the mypy typing errors in tests calling transaction.py to make it easier to validate typing there. I found a few issues already with the tests not always using the correct types at least.

**Comment by janiversen** [comment]
the only reason to mix test files with production code, is if the production code changes needs changes in test !!!

That is an unbreakable rule, just like the CI rule.

However I can just as well tell you, your test changes are unwanted because the only complicate tests and does nothing to improve the production code.

**Comment by jameshilliard** [comment]
> However I can just as well tell you, your test changes are unwanted because the only complicate tests and does nothing to improve the production code.

I'm confused, the test changes are fixing bugs in the tests flagged by mypy. If the mocking in the tests do not use the correct types then the tests are much less useful aren't they?

File: pyproject.toml (hunk 2 above)
Comment (jameshilliard) [reply pyproject.toml]: > Well maybe you have defined the wrong types, which at least seems to be a possibility.

The tests also appear to use the wrong types in a number of places which is why I was trying to fix those in #2249.

> If it's that hard to write tests, then maybe it is also hard to write the correct typing, and not just guessing what is correct !! meaning you could always reduce the PR to the part where you know how to write tests.

It's also hard to use the tests to validate typing issues when there's a lot of [typing issues being flagged by mypy](https://github.com/pymodbus-dev/pymodbus/pull/2249#issuecomment-2241202971) even on `dev`.

> Adding types to code you do not understand, it quite a mountain to climb !! it is hard enough to add typing to code you understand. Just to give you an example, we had discussions during month how to add typing to mixin.py until someone suggested templates.

If we fix the typing errors in the tests it makes it easier to validate that the types here are correct in transaction.py doesn't it?

**Comment by janiversen** [comment]
We do not run mypy and a number of the other tools an test, specifically because there are a larger number of lines flagged, which actually are ok or even needed.

**Comment by janiversen** [comment]
And can we please agree "bugs" are when something do not want as we want, violating types is surely not in that category (and in many cases because we want it).

File: pymodbus/transaction.py [hunk 10]
```diff
@@ -405,23 +407,22 @@ def _transact(self, request: ModbusRequest, response_length, full=False, broadca
...
         """Send."""
         return self.client.framer.sendPacket(packet)
 
-    def _recv(self, expected_response_length, full) -> bytes:  # noqa: C901
+    def _recv(self, expected_response_length: int | None, full: bool) -> bytes:  # noqa: C901
         """Receive."""
         total = None
         if not full:
             exception_length = self._calculate_exception_length()
+            min_size = expected_response_length
```
Comment (janiversen) [inline pymodbus/transaction.py:419]: This change makes no sense, you set the variable just to overwrite it, that is pure waste.

**Review by janiversen — CHANGES_REQUESTED** [review]
There are code changes, which seems untested, and I believe the code changes might be incorrect.

Would be nice to have a test that shows why the old code was wrong, and the new code solves the problem.

File: pymodbus/transaction.py (hunk 10 above)
Comment (jameshilliard) [reply pymodbus/transaction.py:419]: For some reason mypy wasn't liking the previous version here. Not sure what the best approach would be.

File: pymodbus/client/base.py (hunk 9 above)
Comment (jameshilliard) [reply pymodbus/client/base.py:9]: > typecchecking tests is an unwanted feature.

I'm still confused about this, I mean if the tests have incorrect types then they are likely to cause issues.

> Even if we were to consider it, you should NOT mix production code changes with test changes.

Yeah, going to split things out more.

File: .github/workflows/ci.yml [hunk 11]
```diff
@@ -93,7 +93,7 @@ jobs:
       - name: mypy
         if: matrix.run_lint == true
         run: |
-          mypy pymodbus
+          mypy pymodbus test/test_transaction.py test/transport/test_comm.py
```
[the remark this answers is not in the document]
Comment (jameshilliard) [reply .github/workflows/ci.yml]: > Why are you adding ci.yaml when you know that is not the way to make CI changes ?

Was trying to use CI to validate some tests(there was a flaky test that was failing on windows). I removed this now.

File: test/framers/test_old_framers.py [hunk 12]
```diff
@@ -1,4 +1,6 @@
 """Test framers."""
+from __future__ import annotations
```
[the remark this answers is not in the document]
Comment (jameshilliard) [reply test/framers/test_old_framers.py]: removed

**Comment by jameshilliard** [comment]
> We do not run mypy and a number of the other tools an test, specifically because there are a larger number of lines flagged, which actually are ok or even needed.

I'm aware there are currently a lot of mypy errors with the tests...but I see no indication that these aren't fixable by cleaning up the test code in various ways, for example [this](https://github.com/pymodbus-dev/pymodbus/pull/2251) eliminates `method-assign` mypy errors in `test/test_transaction.py` while also eliminating some ugly hacks with how the mocking was being done.

> And can we please agree "bugs" are when something do not want as we want, violating types is surely not in that category (and in many cases because we want it).

I mean, tests using the wrong types are kinda a problem IMO as it makes it more difficult to tell if the production code is correct.

**Comment by alexrudd2** [comment]
> And can we please agree "bugs" are when something do not want as we want, violating types is surely not in that category (and in many cases because we want it).

Here I must disagree.  :)  Although `mypy` has some false positives, making it happy fixes bugs and prevents new ones.  I'd say 20% of the time type errors are actual logic errors, which is high enough to make it worthwhile.

**Comment by janiversen** [comment]
> Here I must disagree.  :)  Although `mypy` has some false positives, making it happy fixes bugs and prevents new ones.  I'd say 20% of the time type errors are actual logic errors, which is high enough to make it worthwhile.

Well if there are actual logic errors, then the code do not work as we want, and thus it is a bug.....problem is that mypy is very do not allow to use the full power of python, being a non typed language, e.g. assigning different types to a variable makes mypy mark it, but it's legal and often in test a nice way to test different parameter types,

**Comment by jameshilliard** [comment]
> Well if there are actual logic errors, then the code do not work as we want, and thus it is a bug.....problem is that mypy is very do not allow to use the full power of python

I mean, I think there's usually ways to make tests play nice with mypy while making the test code cleaner like in #2251 and for cases where it's not possible can just add ignores.

> being a non typed language

Python is both a [strongly typed](https://en.wikipedia.org/wiki/Strong_and_weak_typing) and a [dynamically typed](https://en.wikipedia.org/wiki/Type_system#Dynamic_type_checking_and_runtime_type_information) language, it's certainly not a "non typed language" whatever that means.

> assigning different types to a variable makes mypy mark it, but it's legal and often in test a nice way to test different parameter types,

IMO not being sufficiently strict with types is a major issue that makes the transaction code hard to work with.

**Comment by janiversen** [comment]
> I mean, I think there's usually ways to make tests play nice with mypy while making the test code cleaner like in #2251 and for cases where it's not possible can just add ignores.

I am sure there are.

> Python is both a [strongly typed](https://en.wikipedia.org/wiki/Strong_and_weak_typing) and a [dynamically typed](https://en.wikipedia.org/wiki/Type_system#Dynamic_type_checking_and_runtime_type_information) language, it's certainly not a "non typed language" whatever that means.

It is a non typed language, because you can start using variables without giving them a type, and a variable freely be assigned all kind of types. Python allows typing, but the interpreter do not use that information, which is the definition of a non typed language.



> IMO not being sufficiently strict with types is a major issue that makes the transaction code hard to work with.

Yes but it one of the strong features of python, if you want strict typing then start using e,g, C++

**Comment by jameshilliard** [comment]
> It is a non typed language, because you can start using variables without giving them a type, and a variable freely be assigned all kind of types.

So...dynamically typed I think would be what you're referring to here. Where did the term "no typed language" come from? I'm not seeing much of anything in google about that term.

> Python allows typing, but the interpreter do not use that information, which is the definition of a non typed language.

This is...not accurate at all, what do you think a [TypeError](https://docs.python.org/3/library/exceptions.html#TypeError) is if the interpreter doesn't use typing information?

Python is a strongly typed language, strong typing means that variables do have a type and that the type matters when performing operations on a variable.

> Yes but it one of the strong features of python

I think you're mixing up Python with JavaScript-like languages that are both weakly typed and dynamically typed. Python's type system is more lenient than C++ while still being strongly typed which I prefer to weakly typed languages like JavaScript.

> if you want strict typing then start using e,g, C++

C++ is statically typed with fairly strong typing. Python is dynamically typed with similarly strong typing(I think in some ways even stronger than C++), one reason I like Python is due to it being dynamically and strongly typed. I try to stay away from JavaScript-like languages when possible in general primarily due to them using weak typing which generally results in more bugs.

**Comment by janiversen** [comment]
> So...dynamically typed I think would be what you're referring to here. Where did the term "no typed language" come from? I'm not seeing much of anything in google about that term.


No dynamically typed (e,g, java) is different, as the name indicates a type is added dynamically, and the interpreter uses that to e.g. assign storage.

> This is...not accurate at all, what do you think a [TypeError](https://docs.python.org/3/library/exceptions.html#TypeError) is if the interpreter doesn't use typing information?

For the interpreter everything is stored in a object class, of course within that class there can be types.


> I think you're mixing up Python with JavaScript-like languages that are both weakly typed and dynamically typed. Python's type system is more lenient than C++ while still being strongly typed which I prefer to weakly typed languages like JavaScript.

No I am not ! I have been around from before all these fancy languages, and spent quite a lot of time helping make parsers and interpreters.

Fact is, you can write a perfectly valid python program, without using a single type definition, and if you care to take a look at the python parser, it skips all type information, which is actually currently one of the bigger problems for making an optimizer in coython.

Of course the object class used by the runtime, do checks on assignment etc, but that is due to the internals of the object, not the interpreter,



> C++ is statically typed with fairly strong typing. Python is dynamically typed with similarly strong typing(I think in some ways even stronger than C++), one reason I like Python is due to it being dynamically and strongly typed. I try to stay away from JavaScript-like languages when possible in general primarily due to them using weak typing which generally results in more bugs.

Your sentence is wrong. python allow typing, but do not use that information as it is purely optional.

Easy proof, make an array with 1.000.000 entries with a mixture of integers and floats, measured the actual memory footprint. Then type the array and use only integers, the active memory footprint is the same. A typed language would use the typing information to optimize the memory footprint.

But anyhow believe what you want to believe, I am too old to give free lessons in compiler techniques.

**Comment by jameshilliard** [comment]
> No dynamically typed (e,g, java) is different, as the name indicates a type is added dynamically, and the interpreter uses that to e.g. assign storage.

Java is actually statically typed, JavaScript is dynamically typed at least based on the definitions that I'm familiar with:
["A statically-typed language is a language (such as Java, C, or C++) where variable types are known at compile time. In most of these languages, types must be expressly indicated by the programmer; in other cases (such as OCaml), type inference allows the programmer to not indicate their variable types."](https://developer.mozilla.org/en-US/docs/Glossary/Static_typing)

> For the interpreter everything is stored in a object class, of course within that class there can be types.

Everything may be an object but "there can be types" isn't very precise as ["Every object has an identity, a type and a value."](https://docs.python.org/3/reference/datamodel.html#objects-values-and-types), I think this is somewhat important for making python's strong typing system work.

> Fact is, you can write a perfectly valid python program, without using a single type definition, and if you care to take a look at the python parser, it skips all type information, which is actually currently one of the bigger problems for making an optimizer in coython.

That's true, ["The Python runtime does not enforce function and variable type annotations. They can be used by third party tools such as type checkers, IDEs, linters, etc."](https://docs.python.org/3/library/typing.html). It is of course also true that it's harder to optimize dynamically typed languages like Python.

> Your sentence is wrong. python allow typing, but do not use that information as it is purely optional.

The sort of [typing](https://docs.python.org/3/library/typing.html) I think you are referring to is just optional type annotations, so it is true that typing is optional in that sense. However ["In a dynamically typed language, a variable is simply a value bound to a name; the value has a type -- like "integer" or "string" or "list" -- but the variable itself doesn't."](https://wiki.python.org/moin/Why%20is%20Python%20a%20dynamic%20language%20and%20also%20a%20strongly%20typed%20language).

> Easy proof, make an array with 1.000.000 entries with a mixture of integers and floats, measured the actual memory footprint. Then type the array and use only integers, the active memory footprint is the same. A typed language would use the typing information to optimize the memory footprint.

That's true, ["The Python runtime does not enforce function and variable type annotations."](https://docs.python.org/3/library/typing.html). ["In a dynamically typed language, objects still have a type, but it is determined at runtime."](https://wiki.python.org/moin/Why%20is%20Python%20a%20dynamic%20language%20and%20also%20a%20strongly%20typed%20language)

> But anyhow believe what you want to believe, I am too old to give free lessons in compiler techniques.

I think for the most part you are just using different terminology than what I'm familiar with.

**Comment by janiversen** [comment]
Most people do not care about how the compiler works as long as it does the job, which is a pity because it often leads to programs with a bad performance.

If you solely look at typing from the outside, you really miss a lot, in order to make big scale programs. 

As you correctly noted dynamically typed means a variable is assigned a type at runtime...but you missed the point everything in python is inherited from THE Object so the type of everything in python is Object, This is the reason why you can e,g, assign a string to a function, and thus totally change how it works.

**Comment by janiversen** [comment]
But as I said, stay in your beliefs! 

Fact for this project, typing is accepted because a couple of contributors care about it. Every change that complicates the code, without adding other benefit than pleasing mypy is not accepted.

**Comment by jameshilliard** [comment]
> Most people do not care about how the compiler works as long as it does the job, which is a pity because it often leads to programs with a bad performance.

I am [heavily involved](https://github.com/buildroot/buildroot/commits?author=jameshilliard) with [python toolchain/package integration](https://github.com/buildroot/buildroot/blob/2024.05.1/DEVELOPERS#L1473) as part of the open source [buildroot](https://buildroot.org/) embedded Linux cross-compilation project and have worked on complex cross language toolchain integrations such as buildroot's python rust extension integration(used by libraries like python-cryptography which use [pyo3 bindings](https://github.com/PyO3/pyo3)). So I'm not at all unfamiliar with a lot of lower level details regarding the Python toolchain/interpreter. Python is not a particular easy language to cross compile either but buildroot has excellent support for python cross compilation partially due to a lot of work I've done on it(pep517 build system integrations for example).

> you missed the point everything in python is inherited from THE Object so the type of everything in python is Object, This is the reason why you can e,g, assign a string to a function, and thus totally change how it works.

I don't think I'm disagreeing with that, that's just an implementation detail of python's dynamic typing system.
